File: play-services-core/src/main/java/org/microg/gms/asterism/AsterismService.java
package org.microg.gms.asterism;

import android.os.Binder;
import android.os.Bundle;
import android.os.Parcel;
import android.os.RemoteException;
import android.util.Log;
import com.google.android.gms.common.internal.GetServiceRequest;
import com.google.android.gms.common.internal.IGmsCallbacks;
import org.microg.gms.BaseService;
import org.microg.gms.common.GmsService;

public class AsterismService extends BaseService {
    private static final String TAG = "GmsAsterism";

    public AsterismService() {
        super("GmsAsterism", GmsService.ASTERISM);
    }

    @Override
    public void handleServiceRequest(IGmsCallbacks callback, GetServiceRequest request, GmsService service) throws RemoteException {
        Log.d(TAG, "handleServiceRequest: Interceptando Asterism RCS Handshake");
        // Retorna o Binder que aprova o consentimento do RCS
        callback.onPostInitComplete(0, new AsterismConsentBinder(), null);
    }

    // Binder funcional que responde positivamente ao Google Messages
    private static class AsterismConsentBinder extends Binder {
        @Override
        protected boolean onTransact(int code, Parcel data, Parcel reply, int flags) throws RemoteException {
            if (reply != null) {
                reply.writeNoException();
                Bundle bundle = new Bundle();
                bundle.putInt("consent_status", 1); // 1 = ACCEPTED
                bundle.putBoolean("rcs_eligible", true);
                bundle.putString("terms_version", "v1");
                bundle.writeToParcel(reply, 0);
            }
            return true;
        }
    }
}

File: play-services-core/src/main/kotlin/org/microg/gms/asterism/AsterismService.kt
/**
 * SPDX-FileCopyrightText: 2026 microG Project Team
 * SPDX-License-Identifier: Apache-2.0
 */

package org.microg.gms.asterism

import android.os.Binder
import android.os.Parcel
import android.util.Log
import com.google.android.gms.common.api.CommonStatusCodes
import com.google.android.gms.common.internal.ConnectionInfo
import com.google.android.gms.common.internal.GetServiceRequest
import com.google.android.gms.common.internal.IGmsCallbacks
import org.microg.gms.BaseService
import org.microg.gms.common.GmsService

private const val TAG = "AsterismService"

class AsterismService : BaseService(TAG, GmsService.ASTERISM) {

    override fun handleServiceRequest(callback: IGmsCallbacks, request: GetServiceRequest, service: GmsService) {
        Log.d(TAG, "handleServiceRequest from ${request.callingPackage}")
        callback.onPostInitCompleteWithConnectionInfo(
            CommonStatusCodes.SUCCESS,
            AsterismServiceImpl(),
            ConnectionInfo()
        )
    }
}

class AsterismServiceImpl : Binder() {
    init {
        attachInterface(null, "com.google.android.gms.asterism.internal.IAsterismService")
    }

    override fun onTransact(code: Int, data: Parcel, reply: Parcel?, flags: Int): Boolean {
        Log.d(TAG, "onTransact code: $code")
        reply?.writeNoException()
        return true
    }
}
